e2e-tests.sh

```diff
@@ -29,8 +29,12 @@ function teardown_test_resources() {
   # On boskos, save time and don't teardown as the cluster will be destroyed anyway.
   (( IS_BOSKOS )) && return
   header "Tearing down test environment"
-  function_exists test_teardown && test_teardown
-  function_exists knative_teardown && knative_teardown
+  if function_exists test_teardown; then
```

Comment:
Based on my knowledge on Bash, this should be the same as before.
Could you elaborate what was the bug for L32-33 and how this change fixes it?

Reply:
Oh, that's very different.

```bash
#!/bin/bash

set -Eeuo pipefail

function function_exists() {
  [[ "$(type -t $1)" == "function" ]]
}

function_exists test_teardown && test_teardown

echo "Result: $?"
```

You will get result: `Result: 1`. That's because line `function_exists test_teardown && test_teardown` returns 1. And that's last execution.

When you use the following. The `function_exists test_teardown` is executed as a boolean expression, and the if returns 0, even if the condition is not met (that's the `fi` command BTW).

```bash
if function_exists test_teardown; then
  test_teardown
fi
```

Reply:
Ahh, that's a huge Bash pitfall..
Thanks for explaining!

Reply:
There is plenty of those. Bash only seams obvious.